### PR TITLE
Android Studio / gradle support (WIP)

### PR DESCRIPTION
**Pull request by hrydgard**
EDIT: Updated again. No longer removes the old builds.

### CODE REVIEW

File: android/build.gradle [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+		targetSdkVersion 23
+		ndk {
+			abiFilters 'x86', 'arm64-v8a'
+			// Disabled because of FFMPEG linking issue: 'x86_64'
+			// Disabled because of weird NEON issues: 'armeabi-v7a',
+			// abiFilters 'arm64-v8a'
+		}
+		externalNativeBuild {
+			cmake {
+				// Available arguments are inside ${SDK}/cmake/android.toolchain.cmake file
```
Comment (Orphis) [inline android/build.gradle]: And there too https://developer.android.com/ndk/guides/cmake.html

File: local.properties [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+# Do not modify this file -- YOUR CHANGES WILL BE ERASED!
+#
+# This file must *NOT* be checked into Version Control Systems,
+# as it contains information specific to your local configuration.
+#
+# Location of the SDK. This is only used by Gradle.
+# For customization when using a Version Control System, please read the
+# header note.
+#Sat Oct 08 00:16:27 CEST 2016
+ndk.dir=C\:\\adt\\sdk\\ndk-bundle
```
Comment (Orphis) [inline local.properties]: This file should be git ignored, they're "local" properties!

File: ext/native/ext/cityhash/city.cpp [hunk 3]
```diff
@@ -87,18 +88,19 @@ static uint32 UNALIGNED_LOAD32(const char *p) {
 #endif
 
 #else
+// TODO: Why is this path being built on Android? Slow.
```
Comment (Orphis) [inline ext/native/ext/cityhash/city.cpp]: Because the standard define for Android is actually `__ANDROID__`

File: android/build.gradle [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+				arguments '-DANDROID=true', '-DANDROID_PLATFORM=android-15', '-DANDROID_TOOLCHAIN=clang', '-DANDROID_STL=gnustl_static'
+				cppFlags "-std=c++11"
```
Comment (Orphis) [inline android/build.gradle]: This should be moved in the CMake scripts, not in gradle.

File: CMakeLists.txt [hunk 5]
```diff
@@ -219,38 +233,54 @@ if (NOT CMAKE_BUILD_TYPE)
...
 	add_definitions(-Wno-multichar)
+
+	# Don't compile with strict aliasing, we're not 100% aliasing-safe
 	add_definitions(-fno-strict-aliasing)
 	if (CMAKE_C_COMPILER_ID STREQUAL "Intel")
 		set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -parallel -fopenmp")
 	endif()
 
-	if(X86 AND NOT MIPS)
+	if((X86 OR X86_64) AND NOT MIPS)   # What's with the AND NOT MIPS?
```
Comment (unknownbrackets) [inline CMakeLists.txt]: Probably Loongson is X86 + MIPS?

-[Unknown]

File: CMakeLists.txt [hunk 6]
```diff
@@ -23,16 +33,26 @@ if(CMAKE_SYSTEM_PROCESSOR)
...
+			endif()
 		endif()
-	elseif(${CMAKE_SYSTEM_PROCESSOR} MATCHES "^x86" OR
-		${CMAKE_SYSTEM_PROCESSOR} MATCHES "^amd64" OR
-		${CMAKE_SYSTEM_PROCESSOR} MATCHES "i.86")
+	elseif(${CMAKE_SYSTEM_PROCESSOR} MATCHES "^aarch64")
+	  set(ARM64 ON)
+	  add_definitions(-DARM64)
+	elseif(${CMAKE_SYSTEM_PROCESSOR} MATCHES "^amd64" OR ${CMAKE_SYSTEM_PROCESSOR} MATCHES "^x86_64")
+    set(X86_64 ON)
```
Comment (unknownbrackets) [inline CMakeLists.txt]: nit: A new indentation crime in this indentation war-stricken file...

Maybe we need a `[CMakeLists.txt]` rule in editorconfig?  There's vim support:
https://github.com/editorconfig/editorconfig-vim

-[Unknown]

File: CMakeLists.txt [hunk 7]
```diff
@@ -284,19 +314,20 @@ if(NOT MSVC)
...
 	set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -D_DEBUG")
 	set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -D_NDEBUG")
 endif()
 
-set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib")
-set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "${CMAKE_BINARY_DIR}/lib")
-
 if(ANDROID)
-	set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "${CMAKE_SOURCE_DIR}/android/libs/${ANDROID_ABI}")
+	#set(CMAKE_LIBRARY_OUTPUT_DIRECTORY "${CMAKE_SOURCE_DIR}/android/libs/${ANDROID_ABI}")
```
Comment (unknownbrackets) [inline CMakeLists.txt]: Should we just remove / switch to `NOT ANDROID`?

-[Unknown]

File: CMakeLists.txt [hunk 8]
```diff
@@ -427,13 +476,7 @@ else()
...
 endif()
 
-find_package(Snappy)
-if(SNAPPY_FOUND)
-	add_definitions(-DSHARED_SNAPPY)
-else()
-	add_subdirectory(ext/snappy)
-endif()
-
+add_subdirectory(ext/snappy)
```
Comment (unknownbrackets) [inline CMakeLists.txt]: Will people be mad that we're removing system Snappy support?

-[Unknown]

File: CMakeLists.txt [hunk 9]
```diff
@@ -1500,10 +1560,20 @@ add_library(${CoreLibName} ${CoreLinkType}
...
-target_link_libraries(${CoreLibName} Common native kirk cityhash sfmt19937 xbrz xxhash
+
+if(ANDROID)
+  set(CoreExtraLibs ${CoreExtraLibs} android)
+  if(X86_64)
+    set(CoreExtraLibs -Wl,--gc-sections -Wl,--exclude-libs,ALL)
+  endif()
+endif()	
+
+target_link_libraries(${CoreLibName} Common native kirk cityhash sfmt19937 xbrz xxhash glslang
```
Comment (unknownbrackets) [inline CMakeLists.txt]: Hmm, should `glslang` only be on Vulkan?  Probably it's safe to add it everywhere?

-[Unknown]

File: CMakeLists.txt [hunk 10]
```diff
@@ -1500,10 +1560,20 @@ add_library(${CoreLibName} ${CoreLinkType}
 	$<TARGET_OBJECTS:GPU>
 	Globals.h
 	git-version.cpp)
-target_link_libraries(${CoreLibName} Common native kirk cityhash sfmt19937 xbrz xxhash
+
+if(ANDROID)
+  set(CoreExtraLibs ${CoreExtraLibs} android)
+  if(X86_64)
```
Comment (unknownbrackets) [inline CMakeLists.txt]: Maybe we should add a comment above this explaining why?

-[Unknown]

File: Core/MIPS/MIPSAsm.cpp [hunk 11]
```diff
@@ -7,7 +7,13 @@
 
 #include "Common/CommonTypes.h"
 
-#if defined(_WIN32) || defined(ANDROID)
+// Temporarily turned off on Android
+#if defined(_WIN32)
+#define USE_ARMIPS
```
Comment (unknownbrackets) [inline Core/MIPS/MIPSAsm.cpp]: Just noting: we will want this back for Unittest to work.

-[Unknown]

File: com.bda.controller/build.gradle [hunk 12]
```diff
@@ -0,0 +1,2 @@
+configurations.maybeCreate("default")
```
Comment (unknownbrackets) [inline com.bda.controller/build.gradle]: Bah, it's a shame this has to go at the root - we can't at least shove it inside android/?

-[Unknown]

File: ext/native/ext/cityhash/city.cpp (hunk 3 above)
Comment (unknownbrackets) [inline ext/native/ext/cityhash/city.cpp]: Should we remove this comment now (I think the `|| defined(ANDROID)` is newer?)  By the by, was it using the wrong path before too?  I guess yes?

-[Unknown]

Comment (Orphis) [inline ext/native/ext/cityhash/city.cpp]: Yes comment should be removed. Also I have #9053 to fix all the ANDROID preprocessor defines to __ANDROID__.

File: android/build.gradle (hunk 4 above)
Comment (Orphis) [inline android/build.gradle]: Because of #9053, you can remove the -DANDROID=true part

File: CMakeLists.txt (hunk 10 above)
Comment (Orphis) [inline CMakeLists.txt]: Probably not, except Debian zealots.

**Comment by mrcmunir** [comment]
ext/glslang android.mk miss remove right? :+1:

File: com.bda.controller/build.gradle (hunk 12 above)
Comment (hrydgard) [inline com.bda.controller/build.gradle]: Not quite sure how..

File: CMakeLists.txt (hunk 9 above)
Comment (hrydgard) [inline CMakeLists.txt]: Should be safe.

File: CMakeLists.txt (hunk 5 above)
Comment (hrydgard) [inline CMakeLists.txt]: I don't think it exposes both at the same time..

**Comment by mrcmunir** [comment]
androidjava.io.FileNotFoundException: /home/travis/.android/analytics.settings (No such file or directory)

https://code.google.com/p/android/issues/detail?id=223424

This has been fixed for Android Studio 2.3

A workaround for 2.2 is to set the environment variable "ANDROID_SDK_HOME" to point to a writable directory

**Comment by hrydgard** [comment]
I don't see that, but even worse, travis (or rather gradle) is complaining about license agreements not being accepted:

> You have not accepted the license agreements of the following SDK components:
  [Android SDK Platform 24, Android SDK Build-Tools 23.0.2].
  Before building your project, you need to accept the license agreements and complete the installation of the missing components using the Android Studio SDK Manager.
  Alternatively, to learn how to transfer the license agreements from one workstation to another, go to http://d.android.com/r/studio-ui/export-licenses.html

Silly. Should we simply check in some agreed license files and copy them to the right location? :(

**Comment by hrydgard** [comment]
Hm, here's a diff I found on StackOverflow:

https://github.com/Klauswk/FinancialControl/commit/0f588d072863ee8046d27c090c9ed03b69fe5fdc

There are license components?

**Comment by unknownbrackets** [comment]
It looks like we can, yes: http://stackoverflow.com/questions/38096225/automatically-accept-all-sdk-licences

Also, I saw the analytics error too, apparently `mkdir /home/travis/.android`.  Sorry, I didn't get a chance to try it - may tomorrow.

-[Unknown]

**Comment by unknownbrackets** [comment]
Generally seems worse in a lot of ways for NDK, and won't let me configure number of jobs, and gradle won't output the compile output.  `ant` was easy and did everything sanely, but it's old and being replaced - I get it.

I haven't really had as much time lately, I guess this is a sign I probably shouldn't waste my time anymore.  What great progress in build systems: everything is worse.  Orphis suggests I keep a dirty commit in all my branches forever more to workaround.  I guess new build systems don't even know how to integrate with version control properly anymore.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, it doesn't really seem like an improvement .. only thing is that it's be nice to unify on CMake, also the debugger integration is better.

Totally understand that you don't want to waste your time messing with this ...

**Comment by unknownbrackets** [comment]
Well, I did get it to build before making that comment, but it has been a generally frustrating experience:

https://github.com/hrydgard/ppsspp/commit/ab46fb4b9041cc8e8382b6f288b645239442a230

With `--debug` it will (after completing the build, not real time) output the build log, but it outputs so much debug info that Travis won't show it anyway.  `--info` is not enough.  On my machine, I'm now forced to let it use all my cores and make other apps slow - so I have to sit on my thumbs while it builds, which is high on the list of things I hate most.  If a build takes several minutes, at least allow me to use my computer for other things while it's building and stay productive/entertained/something.

All this so that I can regain the breakpoints I already had with ARM DS-5...

-[Unknown]

**Comment by hrydgard** [comment]
Yeah :(

If we can't get this to work better than it is, we might have to delay merging until Google gets its act together with gradle. I have no idea why it feels so slow and heavy.

**Comment by hrydgard** [comment]
Also it must be possible to force it to use less cores by passing some argument along to ninja. Ninja does take a -j flag just like make, it's just that it defaults to maxing your cores, which is indeed quite annoying.

Anyway, if we decide not to merge this, I'm just gonna maintain this branch sporadically, and use it when I really need an Android debugger.

**Comment by Orphis** [comment]
I don't think we can pass arguments to CMake at all.
But we can use this to limit the amount of concurrent jobs: https://cmake.org/cmake/help/v3.6/prop_gbl/JOB_POOLS.html
Then, we can look for a file like "local.properties", read it and extract a value to set up a limit. It can be optional, doesn't need a dirty tree or rebasing constantly. That should be enough to make it work for you, are you ok with a solution like that?

In the mean time, we can still get this to work and merge the Gradle part. We could keep the old ant build system around too until Google fixes the major issues we're facing.
What do you think?

**Comment by hrydgard** [comment]
Hm, that's true, it wouldn't take very many changes to also keep Ant around, though that would mean still updating Android.mk by hand, but we already do that...

**Comment by Orphis** [comment]
Another possibility is also to make ant use CMake...

**Comment by Orphis** [comment]
Information I got from #android-dev on Freenode:
The toolchain is in $NDK/build/cmake from NDK r13. So we can use a recent CMake and point to it directly, no need to tweak anything in the SDK.
People seems to be happy with Gradle. But not with the plugin, so instead they have an "exec" task that is running CMake and building. I guess we could just do that.

**Comment by hrydgard** [comment]
I think we'll simply postpone this until it works better and more reliably (maybe a few Android Studio releases from now....)

That also means that we can stick with CMake 3.5 for now.

**Comment by Orphis** [comment]
Postponing this change, sure.
But we want to use a recent NDK and for that, we need Google's CMake toolchain file (that is both in r13 and the CMake SDK manager distribution) and that requires 3.6.
The plan for Android is to either use Ant or Gradle for the Java part, but build the native part with CMake somehow.

**Comment by hrydgard** [comment]
The annoying issue of not being able to install CMake properly from the command line is being fixed soon, at least:

https://code.google.com/p/android/issues/detail?id=221907

**Comment by Orphis** [comment]
New NDK r14 and Gradle buildtools have been released! Time to update this.

**Comment by InusualZ** [comment]
Gradle and Android Studio are not strictly necessary. CMake added support for Android since 3.7. You can build the entire application without them. Although you would have to "upgrade" your CMake version

**Comment by hrydgard** [comment]
@InusualZ  got any instructions for that (also building the Java part using cmake) ?

**Comment by InusualZ** [comment]
The java part I'm not entirely sure but I think so. if not I think there was an "plugin" companion for AndroidCMake that could do it. About the instruction, the 3.7 documentation is pretty good. But not that I remember. In window you would need Ninja Build as replacement for Make. So you may think it's not worth it

**Comment by Orphis** [comment]
You can't build the Java part with it, so it's a dead end. 

Using Gradle has some advantages : you get a working dev environment right away when using Android Studio (which is what most people will use). 
Upgrading CMake isn't a problem, we already use 3.7 in CI anyway.

**Comment by hrydgard** [comment]
Re-did this again.

Like before, it does not remove the possibility to do ant/eclipse builds and similar, instead it provides an alternative - useful when you want to use the Android Studio debugger or build from the command line with non-deprecated tools (again, ant will keep working, it's just "officially" deprecated by google). So I intend to merge this soon.

File: .gitignore [hunk 13]
```diff
@@ -47,20 +47,18 @@ PPSSPPControls.dat
...
 Logs
 Memstick
 memstick
 Cheats
 
-bin
-gen
-libs
-obj
-build*/
```
Comment (unknownbrackets) [inline .gitignore]: We've historically been telling people to create dirs like `build-ios` and `build-mac` on the root, probably better to keep this one?  Could prefix it with `/`.

-[Unknown]

Comment (Orphis) [inline .gitignore]: Yes please, keep build*/

Comment (hrydgard) [inline .gitignore]: Done!

**Pull request merged**